test/shared/useTask.test.ts

```diff
@@ -0,0 +1,163 @@
+import test, { Test } from 'tape';
+import AbortController from 'abort-controller';
+import { useTask, Task } from '@/index';
+
+test('it can succeed with an asynchronous callback', async (t: Test) => {
+  // Given a "pending" async task that returns a number.
+  const task = useTask(async () => {
+    t.equal(task.getStatus(), 'running');
+    return 42;
+  });
+  t.equal(task.getResult(), undefined);
+  t.equal(task.getStatus(), 'pending');
+
+  // When we run the task.
+  await task.run();
+
+  // Then we get the right result and it was marked as successful.
+  t.equal(task.getResult(), 42);
+  t.equal(task.getStatus(), 'successful');
+  t.equal(task.getError(), undefined);
+});
+
+test('it can succeed with an synchronous callback', async (t: Test) => {
+  // Given a "pending" task that returns a number.
+  const task = useTask(() => {
+    t.equal(task.getStatus(), 'running');
+    return 42;
+  });
+  t.equal(task.getResult(), undefined);
+  t.equal(task.getStatus(), 'pending');
+
+  // When we run the task.
+  await task.run();
+
+  // Then we get the right result and it was marked as successful.
+  t.equal(task.getResult(), 42);
+  t.equal(task.getStatus(), 'successful');
+  t.equal(task.getError(), undefined);
+});
+
+test('it can fail', async (t: Test) => {
+  // Given a "pending" task that throws an error.
+  const task = useTask(() => {
+    t.equal(task.getStatus(), 'running');
+    throw new Error('Test Task Failure');
+  });
+  t.equal(task.getResult(), undefined);
+  t.equal(task.getStatus(), 'pending');
+
+  // When we run the task.
+  try {
+    await task.run();
+  } catch (error) {
+    // Then the task is marked as failed and we kept track of the error.
+    t.equal(task.getStatus(), 'failed');
+    t.equal(task.getResult(), undefined);
+    t.equal(task.getError(), error);
+    return;
+  }
+
+  t.fail('Test task should have failed');
```

Comment:
I'd remove the `return` from the `catch` block and move this line into the `try` block (as the last line) as it should fail before it reaches it.

Reply:
Agreed!